plugins/analogix/fu-analogix-device.c

```diff
@@ -0,0 +1,503 @@
+	    hdr->custom_payload_len > 0) {
+		base = hdr->fw_payload_len + hdr->secure_tx_payload_len + hdr->secure_rx_payload_len;
+		if (!fu_analogix_device_program_flash (self,
+						       hdr->custom_start_addr,
+						       hdr->total_len,
+						       hdr->custom_payload_len,
+						       ANX_BB_WVAL_UPDATE_CUSTOM_DEF,
+						       base,
+						       fw_payload,
+						       error)) {
+			g_prefix_error (error, "program custom define failed: ");
+			return FALSE;
+		}
+	}
+	if (hdr->secure_tx_start_addr == FLASH_TXFW_ADDR &&
+	    hdr->secure_tx_payload_len > 0) {
+		base = hdr->fw_payload_len;
+		if (!fu_analogix_device_program_flash (self,
+						       hdr->secure_tx_start_addr,
+						       hdr->total_len,
+						       hdr->secure_tx_payload_len,
+						       ANX_BB_WVAL_UPDATE_SECURE_TX,
+						       base,
+						       fw_payload,
+						       error)) {
+			g_prefix_error (error, "program secure OCM TX failed: ");
+			return FALSE;
+		}
+	}
+	if (hdr->secure_rx_start_addr == FLASH_RXFW_ADDR &&
+	    hdr->secure_rx_payload_len > 0) {
+		base = hdr->fw_payload_len + hdr->secure_tx_payload_len;
+		if (!fu_analogix_device_program_flash (self,
+						       hdr->secure_rx_start_addr,
+						       hdr->total_len,
+						       hdr->secure_rx_payload_len,
+						       ANX_BB_WVAL_UPDATE_SECURE_RX,
+						       base,
+						       fw_payload,
+						       error)) {
+			g_prefix_error (error, "program secure OCM RX failed: ");
+			return FALSE;
+		}
+	}
+	if (hdr->fw_start_addr == FLASH_OCM_ADDR && hdr->fw_payload_len > 0) {
+		base = 0;
+		if (!fu_analogix_device_program_flash (self,
+						       hdr->fw_start_addr,
+						       hdr->total_len,
+						       hdr->fw_payload_len,
+						       ANX_BB_WVAL_UPDATE_OCM,
+						       base,
+						       fw_payload,
+						       error)) {
+			g_prefix_error (error, "program OCM failed: ");
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_analogix_device_close (FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	FuAnalogixDevice *self = FU_ANALOGIX_DEVICE (device);
+
+	/* release interface */
+	if (!g_usb_device_release_interface (usb_device, self->iface_idx,
+					     G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					     error)) {
+		g_prefix_error (error, "failed to release interface: ");
+		return FALSE;
+	}
+
+	/* FuUsbDevice->close */
+	return FU_DEVICE_CLASS (fu_analogix_device_parent_class)->close (device, error);
+}
+
+static void
+fu_analogix_device_init (FuAnalogixDevice *self)
+{
+	fu_device_add_protocol (FU_DEVICE (self), "com.analogix.bb");
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
+	fu_device_set_version_format (FU_DEVICE (self), FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_set_summary (FU_DEVICE (self), "Phoenix-Lite");
```

Comment:
Shouldn't this be in the quirk file, as presumably the protocol might be used by multiple devices eventually?

Reply:
Removed.